In [1]:
#Importing the Python Libraries

import numpy as np
import scipy as sp
import pandas as pd
import matplotlib as mpl
import seaborn as sns
from IPython.display import display, HTML
import warnings

#Ignore warnings
warnings.filterwarnings('ignore')

In [2]:
# Loading the Dataset
DataFrame = pd.read_csv('C:/Users/ishan/Downloads/malware_BinaryImbalanced.csv')

# Printing and Displaying the Dataframe
DataFrame.head(10)

hash  millisecond classification  \
0  com.kmcpesh.medicalskillsproceduresfree.apk            0         benign   
1  com.kmcpesh.medicalskillsproceduresfree.apk            1         benign   
2  com.kmcpesh.medicalskillsproceduresfree.apk            2         benign   
3  com.kmcpesh.medicalskillsproceduresfree.apk            3         benign   
4  com.kmcpesh.medicalskillsproceduresfree.apk            4         benign   
5  com.kmcpesh.medicalskillsproceduresfree.apk            5         benign   
6  com.kmcpesh.medicalskillsproceduresfree.apk            6         benign   
7  com.kmcpesh.medicalskillsproceduresfree.apk            7         benign   
8  com.kmcpesh.medicalskillsproceduresfree.apk            8         benign   
9  com.kmcpesh.medicalskillsproceduresfree.apk            9         benign   

        os  state  usage_counter        prio  static_prio  normal_prio  \
0   Ubuntu      0              0  3069403136        16447            0   
1   CentOS      0              0  3069403136        16447            0   
2   Ubuntu      0              0  3069403136        16447            0   
3   CentOS      0              0  3069403136        16447            0   
4      Mac      0              0  3069403136        16447            0   
5  Windows      0              0  3069403136        16447            0   
6  Windows      0              0  3069403136        16447            0   
7   CentOS      0              0  3069403136        16447            0   
8   Ubuntu      0              0  3069403136        16447            0   
9      Mac      0              0  3069403136        16447            0   

   policy  ...  nivcsw  min_flt  maj_flt  fs_excl_counter        lock   utime  \
0       0  ...       0        1      120                0  3204448256  384502   
1       0  ...       0        1      120                0  3204448256  384502   
2       0  ...       0        1      120                0  3204448256  384502   
3       0  ...       0        1      120                0  3204448256  384502   
4       0  ...       0        1      120                0  3204448256  384502   
5       0  ...       0        1      120                0  3204448256  384502   
6       0  ...       0        1      120                0  3204448256  384502   
7       0  ...       0        1      120                0  3204448256  384502   
8       0  ...       0        1      120                0  3204448256  384502   
9       0  ...       0        1      120                0  3204448256  384502   

   stime  gtime  cgtime  signal_nvcsw  
0      5      0       0             0  
1      5      0       0             0  
2      5      0       0             0  
3      5      0       0             0  
4      5      0       0             0  
5      5      0       0             0  
6      5      0       0             0  
7      5      0       0             0  
8      5      0       0             0  
9      5      0       0             0  

[10 rows x 36 columns]

In [3]:
# Displaying the shape of the DataFrame
print("Shape of the DataFrame:", DataFrame.shape)

# Selecting columns which are relevant for further analysis on given 19 columns
selected_columns = ['classification','os','usage_counter','prio','static_prio','normal_prio','vm_pgoff','vm_truncate_count','task_size','map_count','hiwater_rss','total_vm','shared_vm','exec_vm','reserved_vm','nr_ptes','nvcsw','nivcsw','signal_nvcsw']

# Creating the DataFrame for the selected columns
DataFrame_org = DataFrame[selected_columns]

# Displaying the rows of the selected DataFrame
DataFrame_org.head(5)

Shape of the DataFrame: (100000, 36)


classification      os  usage_counter        prio  static_prio  normal_prio  \
0         benign  Ubuntu              0  3069403136        16447            0   
1         benign  CentOS              0  3069403136        16447            0   
2         benign  Ubuntu              0  3069403136        16447            0   
3         benign  CentOS              0  3069403136        16447            0   
4         benign     Mac              0  3069403136        16447            0   

   vm_pgoff  vm_truncate_count  task_size  map_count  hiwater_rss  total_vm  \
0         0              14739          0       7903            0        88   
1         0              14739          0       7903            0        88   
2         0              14739          0       7903            0        88   
3         0              14739          0       7903            0        88   
4         0              14739          0       7903            0        88   

   shared_vm  exec_vm  reserved_vm  nr_ptes   nvcsw  nivcsw  signal_nvcsw  
0        120      120           80        0  349169       0             0  
1        120      120           80        0  349169       0             0  
2        120      120           80        0  349169       0             0  
3        120      120           80        0  349169       0             0  
4        120      120           80        0  349169       0             0

In [4]:
# Checking for missing values in DataFrame
print('\nColumnName, DataType, MissingValues')
for col in DataFrame.columns:
    print(col, ',', DataFrame[col].dtype, ',', DataFrame[col].isnull().any())


ColumnName, DataType, MissingValues
hash , object , False
millisecond , int64 , False
classification , object , False
os , object , False
state , int64 , False
usage_counter , int64 , False
prio , int64 , False
static_prio , int64 , False
normal_prio , int64 , False
policy , int64 , False
vm_pgoff , int64 , False
vm_truncate_count , int64 , False
task_size , int64 , False
cached_hole_size , int64 , False
free_area_cache , int64 , False
mm_users , int64 , False
map_count , int64 , False
hiwater_rss , int64 , False
total_vm , int64 , False
shared_vm , int64 , False
exec_vm , int64 , False
reserved_vm , int64 , False
nr_ptes , int64 , False
end_data , int64 , False
last_interval , int64 , False
nvcsw , int64 , False
nivcsw , int64 , False
min_flt , int64 , False
maj_flt , int64 , False
fs_excl_counter , int64 , False
lock , int64 , False
utime , int64 , False
stime , int64 , False
gtime , int64 , False
cgtime , int64 , False
signal_nvcsw , int64 , False


In [5]:
# Creating a copy of the original DataFrame
DataFrame_copy=DataFrame_org.copy(deep=True)
print(DataFrame_copy.columns)
DataFrame_copy.head(5)

Index(['classification', 'os', 'usage_counter', 'prio', 'static_prio',
       'normal_prio', 'vm_pgoff', 'vm_truncate_count', 'task_size',
       'map_count', 'hiwater_rss', 'total_vm', 'shared_vm', 'exec_vm',
       'reserved_vm', 'nr_ptes', 'nvcsw', 'nivcsw', 'signal_nvcsw'],
      dtype='object')


classification      os  usage_counter        prio  static_prio  normal_prio  \
0         benign  Ubuntu              0  3069403136        16447            0   
1         benign  CentOS              0  3069403136        16447            0   
2         benign  Ubuntu              0  3069403136        16447            0   
3         benign  CentOS              0  3069403136        16447            0   
4         benign     Mac              0  3069403136        16447            0   

   vm_pgoff  vm_truncate_count  task_size  map_count  hiwater_rss  total_vm  \
0         0              14739          0       7903            0        88   
1         0              14739          0       7903            0        88   
2         0              14739          0       7903            0        88   
3         0              14739          0       7903            0        88   
4         0              14739          0       7903            0        88   

   shared_vm  exec_vm  reserved_vm  nr_ptes   nvcsw  nivcsw  signal_nvcsw  
0        120      120           80        0  349169       0             0  
1        120      120           80        0  349169       0             0  
2        120      120           80        0  349169       0             0  
3        120      120           80        0  349169       0             0  
4        120      120           80        0  349169       0             0

In [6]:
from sklearn import preprocessing
from IPython.display import display, HTML

# Encoding the labels

# Defining the label as nominal values
y = DataFrame_org['classification'] 
le = preprocessing.LabelEncoder()
le.fit(y)

# Encoding nominal labels to integers
y_encoded = le.transform(y)  
DataFrame_org['classification'] = y_encoded

DataFrame_org.head(5)

classification      os  usage_counter        prio  static_prio  \
0               0  Ubuntu              0  3069403136        16447   
1               0  CentOS              0  3069403136        16447   
2               0  Ubuntu              0  3069403136        16447   
3               0  CentOS              0  3069403136        16447   
4               0     Mac              0  3069403136        16447   

   normal_prio  vm_pgoff  vm_truncate_count  task_size  map_count  \
0            0         0              14739          0       7903   
1            0         0              14739          0       7903   
2            0         0              14739          0       7903   
3            0         0              14739          0       7903   
4            0         0              14739          0       7903   

   hiwater_rss  total_vm  shared_vm  exec_vm  reserved_vm  nr_ptes   nvcsw  \
0            0        88        120      120           80        0  349169   
1            0        88        120      120           80        0  349169   
2            0        88        120      120           80        0  349169   
3            0        88        120      120           80        0  349169   
4            0        88        120      120           80        0  349169   

   nivcsw  signal_nvcsw  
0       0             0  
1       0             0  
2       0             0  
3       0             0  
4       0             0

In [7]:
# Data preprocessing

print('Column DataFrametypes:\n',DataFrame_org.dtypes)

# Converting all the nominal variables to binary variables
DataFrame_num=DataFrame_org.copy(deep=True)

# Creating the new binary columns
DataFrame_dummies=pd.get_dummies(DataFrame_num[['os']], dtype= int)

# Merging the DataFrames
DataFrame_num=DataFrame_num.join(DataFrame_dummies)

# Droping the original columns
DataFrame_num=DataFrame_num.drop('os',axis=1)
DataFrame_num=DataFrame_num.drop('os_Windows', axis=1)

DataFrame_num.head(5)


Column DataFrametypes:
 classification        int32
os                   object
usage_counter         int64
prio                  int64
static_prio           int64
normal_prio           int64
vm_pgoff              int64
vm_truncate_count     int64
task_size             int64
map_count             int64
hiwater_rss           int64
total_vm              int64
shared_vm             int64
exec_vm               int64
reserved_vm           int64
nr_ptes               int64
nvcsw                 int64
nivcsw                int64
signal_nvcsw          int64
dtype: object


classification  usage_counter        prio  static_prio  normal_prio  \
0               0              0  3069403136        16447            0   
1               0              0  3069403136        16447            0   
2               0              0  3069403136        16447            0   
3               0              0  3069403136        16447            0   
4               0              0  3069403136        16447            0   

   vm_pgoff  vm_truncate_count  task_size  map_count  hiwater_rss  ...  \
0         0              14739          0       7903            0  ...   
1         0              14739          0       7903            0  ...   
2         0              14739          0       7903            0  ...   
3         0              14739          0       7903            0  ...   
4         0              14739          0       7903            0  ...   

   exec_vm  reserved_vm  nr_ptes   nvcsw  nivcsw  signal_nvcsw  os_CentOS  \
0      120           80        0  349169       0             0          0   
1      120           80        0  349169       0             0          1   
2      120           80        0  349169       0             0          0   
3      120           80        0  349169       0             0          1   
4      120           80        0  349169       0             0          0   

   os_Debian  os_Mac  os_Ubuntu  
0          0       0          1  
1          0       0          0  
2          0       0          1  
3          0       0          0  
4          0       1          0  

[5 rows x 22 columns]

In [8]:
# N-1 binary column
DataFrame_num.head(5)

classification  usage_counter        prio  static_prio  normal_prio  \
0               0              0  3069403136        16447            0   
1               0              0  3069403136        16447            0   
2               0              0  3069403136        16447            0   
3               0              0  3069403136        16447            0   
4               0              0  3069403136        16447            0   

   vm_pgoff  vm_truncate_count  task_size  map_count  hiwater_rss  ...  \
0         0              14739          0       7903            0  ...   
1         0              14739          0       7903            0  ...   
2         0              14739          0       7903            0  ...   
3         0              14739          0       7903            0  ...   
4         0              14739          0       7903            0  ...   

   exec_vm  reserved_vm  nr_ptes   nvcsw  nivcsw  signal_nvcsw  os_CentOS  \
0      120           80        0  349169       0             0          0   
1      120           80        0  349169       0             0          1   
2      120           80        0  349169       0             0          0   
3      120           80        0  349169       0             0          1   
4      120           80        0  349169       0             0          0   

   os_Debian  os_Mac  os_Ubuntu  
0          0       0          1  
1          0       0          0  
2          0       0          1  
3          0       0          0  
4          0       1          0  

[5 rows x 22 columns]

In [26]:
# Logistic regression

# API
# https://scikit-learn.org/0.16/modules/generated/sklearn.linear_model.LogisticRegression.html

import pandas as pd
import numpy as np
from sklearn import metrics 
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import make_scorer, precision_score, accuracy_score
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score, roc_auc_score

from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix as cm

x=DataFrame_num.drop('classification',axis=1)
y=DataFrame_num['classification']

# By using hold-out evaluation
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)
clfn=LogisticRegression(penalty='l2',solver='newton-cg', 
                       max_iter=150, multi_class='ovr')
clfn=clfn.fit(x_train, y_train)
y_pred=clfn.predict(x_test)

confM = cm(y_test, y_pred)
print(confM)

Accuracy=accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='macro')
Auc_reg = roc_auc_score(y_test, y_pred, average = 'macro')

print('Logistic regression by hold-out evaluation: \nAccuracy = ',Accuracy, ', \nF1 = ', f1, '\nAUC_reg = ', Auc_reg)


[[17537   470]
 [ 1515   478]]
Logistic regression by hold-out evaluation: 
Accuracy =  0.90075 , 
F1 =  0.6357481332848778 
AUC_reg =  0.6068692386478125


In [27]:
# SVM - poly

from sklearn.svm import SVC

# API for SVC
# https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html?highlight=svc#sklearn.svm.SVC

# By using hold-out evaluation

# Spliting
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state= 46)

clfn=SVC(kernel='poly', C=0.1, max_iter=-1) # C is large -> hard margin; C is small -> soft margin. Running time may be longer if C is larger
clfn=clfn.fit(x_train, y_train)
y_pred=clfn.predict(x_test)

#Calculating different metrics
Acc_poly =accuracy_score(y_test, y_pred)
f1_poly = f1_score(y_test, y_pred, average='macro')
Auc_poly = roc_auc_score(y_test,y_pred, average = 'macro')

print('SVM by hold-out evaluation for poly kernel: \nAcc_poly = ',Acc_poly, ', \nF1 = ', f1_poly, '\nAUC_poly = ', Auc_poly)


SVM by hold-out evaluation for poly kernel: 
Acc_poly =  0.89928 , 
F1 =  0.4734846889347542 
AUC_poly =  0.5


In [28]:
# SVM - rbf

from sklearn.svm import SVC

# API for SVC
# https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html?highlight=svc#sklearn.svm.SVC

# By using hold-out evaluation

# Splitting
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state= 42)

clfn=SVC(kernel='rbf', C=0.1, max_iter=-1) # C is large -> hard margin; C is small -> soft margin. Running time may be longer if C is larger
clfn=clfn.fit(x_train, y_train)
y_pred=clfn.predict(x_test)

Acc_rbf =accuracy_score(y_test, y_pred)
f1_rbf = f1_score(y_test, y_pred, average='macro')
Auc_rbf= roc_auc_score(y_test, y_pred, average = 'macro')

print('SVM by hold-out evaluation for rbf kernel: \nAccuracy = ',Acc_rbf, ', \nF1 = ', f1_rbf, '\nAUC_rbf = ', Auc_rbf)


SVM by hold-out evaluation for rbf kernel: 
Accuracy =  0.89932 , 
F1 =  0.4734957774361351 
AUC_rbf =  0.5


In [29]:
# SVM - linear 

from sklearnex import patch_sklearn 
patch_sklearn()
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split  
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score  


# API for SVC
# https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html?highlight=svc#sklearn.svm.SVC

# By using hold-out evaluation

# Splitting
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state= 42)

clfn = SVC(kernel='linear', C=0.1, max_iter=-1, decision_function_shape='ovr')  # C is large -> hard margin; C is small -> soft margin. Running time may be longer if C is larger
clfn = clfn.fit(x_train, y_train)
y_pred = clfn.predict(x_test)

Acc_lin = accuracy_score(y_test, y_pred)
f1_lin = f1_score(y_test, y_pred, average='macro')
Auc_lin = roc_auc_score(y_test, y_pred, average='macro')

print('SVM by hold-out evaluation for linear kernel: \nAccuracy = ', Acc_lin, ', \nF1 = ', f1_lin, '\nAUC_lin = ', Auc_lin)


Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)


SVM by hold-out evaluation for linear kernel: 
Accuracy =  0.76064 , 
F1 =  0.6243628191546317 
AUC_lin =  0.7720130078063131


In [30]:
# RandomForest

# API
# https://scikit-learn.org/0.16/modules/generated/sklearn.linear_model.LogisticRegression.html

from sklearn import metrics 
from sklearn.model_selection import train_test_split
from sklearn.metrics import make_scorer, precision_score, accuracy_score
from sklearn.model_selection import cross_val_score

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier

# https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html?highlight=decisiontreeclassifier#sklearn.tree.DecisionTreeClassifier
# https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html

x=DataFrame_num.drop('classification',axis=1)
y=DataFrame_num['classification']

# By using hold-out evaluation

# Splitting
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state= 46)

tree = DecisionTreeClassifier()
bag = BaggingClassifier(tree, n_estimators=100, max_samples=0.85, random_state=1)
clfn=bag.fit(x_train, y_train)
y_pred=clfn.predict(x_test)

Acc_rf =accuracy_score(y_test, y_pred)
f1_rf = f1_score(y_test, y_pred, average='macro')
Auc_rf= roc_auc_score(y_test, y_pred, average = 'macro')

print('Random forest by hold-out evaluation: \nAccuracy = ',Acc_rf, ', \nF1 = ', f1_rf, '\nAUC_rf = ', Auc_rf)


Random forest by hold-out evaluation: 
Accuracy =  1.0 , 
F1 =  1.0 
AUC_rf =  1.0


In [31]:
# RandomForest

# API
# https://scikit-learn.org/0.16/modules/generated/sklearn.linear_model.LogisticRegression.html

from sklearn import metrics 
from sklearn.model_selection import train_test_split
from sklearn.metrics import make_scorer, precision_score, accuracy_score
from sklearn.model_selection import cross_val_score

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier

# https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html?highlight=decisiontreeclassifier#sklearn.tree.DecisionTreeClassifier
# https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html

x=DataFrame_num.drop('classification',axis=1)
y=DataFrame_num['classification']

# By using hold-out evaluation

# Splitting
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state= 46)

tree = DecisionTreeClassifier()
bag = BaggingClassifier(tree, n_estimators=500, max_samples=0.85, random_state=1)
clfn=bag.fit(x_train, y_train)
y_pred=clfn.predict(x_test)

Acc_rf =accuracy_score(y_test, y_pred)
f1_rf = f1_score(y_test, y_pred, average='macro')
Auc_rf= roc_auc_score(y_test, y_pred, average = 'macro')

print('Random forest by hold-out evaluation: \nAccuracy = ',Acc_rf, ', \nF1 = ', f1_rf, '\nAUC_pp = ', Auc_rf)


Random forest by hold-out evaluation: 
Accuracy =  0.99996 , 
F1 =  0.9998895753742483 
AUC_pp =  0.999801429706116


In [32]:
# Bagging (USING NAIVE BAYES AS CLASSIFIER) (Bernoulli NB)

from sklearn.naive_bayes import BernoulliNB
from sklearn.ensemble import BaggingClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score

# By using hold-out evaluation

# Splitting
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=46)

# Example of randomForest = bagging method of decision trees
clfn = BernoulliNB()
bag = BaggingClassifier(clfn, n_estimators=100, max_samples=0.85, random_state=1)
clfn = bag.fit(x_train, y_train)
y_pred = clfn.predict(x_test)

Acc_bern = accuracy_score(y_test, y_pred)
f1_bern = f1_score(y_test, y_pred, average='macro')
Auc_bern = roc_auc_score(y_test, y_pred, average='macro')

print('Bagging by hold-out evaluation for Bernoulli NB: \nAccuracy = ', Acc_bern, ', \nF1 = ', f1_bern, '\nAUC_bern = ', Auc_bern)


Bagging by hold-out evaluation for Bernoulli NB: 
Accuracy =  0.89928 , 
F1 =  0.4734846889347542 
AUC_bern =  0.5


In [33]:
# Gradient Boosting Classifier

from sklearn.ensemble import GradientBoostingClassifier
# API, https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html

# By using hold-out evaluation

# Splitting
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)

# Initialize the GradientBoostingClassifier
clfn = GradientBoostingClassifier(n_estimators=100, random_state=0, learning_rate=0.1, criterion='squared_error')
clfn.fit(X_train, y_train)
y_pred = clfn.predict(X_test)

# Calculating the metrics
Acc_grad = accuracy_score(y_test, y_pred)
f1_grad = f1_score(y_test, y_pred, average='macro')
Auc_grad = roc_auc_score(y_test, y_pred, average = 'macro')

print('Gradient Boosting Classifier by hold-out evaluation: \nAccuracy = ',Acc_grad, ', \nF1 = ', f1_grad, '\nAUC_grad = ', Auc_grad)


Gradient Boosting Classifier by hold-out evaluation: 
Accuracy =  0.99948 , 
F1 =  0.9985609252805883 
AUC_grad =  0.9974175605880016


In [34]:
# Gradient Boosting Classifier

from sklearn.ensemble import GradientBoostingClassifier
# API, https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html

# By using hold-out evaluation

# Splitting
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)

# Initialize the GradientBoostingClassifier
clfn = GradientBoostingClassifier(n_estimators=100, random_state=0, learning_rate=0.5, criterion='squared_error')
clfn.fit(X_train, y_train)
y_pred = clfn.predict(X_test)

# Calculating the metrics
Acc_grad = accuracy_score(y_test, y_pred)
f1_grad = f1_score(y_test, y_pred, average='macro')
Auc_grad = roc_auc_score(y_test, y_pred, average = 'macro')

print('Gradient Boosting Classifier by hold-out evaluation: \nAccuracy = ',Acc_grad, ', \nF1 = ', f1_grad, '\nAUC_grad = ', Auc_grad)


Gradient Boosting Classifier by hold-out evaluation: 
Accuracy =  1.0 , 
F1 =  1.0 
AUC_grad =  1.0


In [39]:
# AdaBoosting

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score
from sklearn.model_selection import train_test_split

tree = DecisionTreeClassifier(criterion='gini', ccp_alpha= 1.0)
adaboost_params = { 'base_estimator': tree, 'n_estimators': 100, 'random_state': 0}

clfn = AdaBoostClassifier(tree, n_estimators=100, random_state=1, algorithm='SAMME')

# Splitting the data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=0)

clfn.fit(x_train, y_train)
y_pred = clfn.predict(x_test)

# Calculate Accuracy, F1 Score, and AUC
Acc = accuracy_score(y_test, y_pred)
F1 = f1_score(y_test, y_pred, average='macro')
AUC = roc_auc_score(y_test, y_pred, average='macro')

print("AdaBoosting by Hold-Out Evaluation: \nAccuracy = ", Acc, ", \nF1 = ", F1, ", \nAUC = ", AUC)


AdaBoosting by Hold-Out Evaluation: 
Accuracy =  0.89868 , 
F1 =  0.4733183053489793 , 
AUC =  0.5


In [40]:
# XGBoost

from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score
from sklearn.model_selection import train_test_split

clfn = XGBClassifier( learning_rate=0, n_estimators=100, max_depth=3, min_child_weight=1, gamma=0)

# Splitting the data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=0)

clfn.fit(x_train, y_train)
y_pred = clfn.predict(x_test)

# Calculate Accuracy, F1 Score, and AUC
Acc_xgb = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='macro')
Auc_xgb = roc_auc_score(y_test, y_pred, average='macro')

print("XGBoost by Hold-Out Evaluation: \nAccuracy = ", Acc_xgb, ", \nF1 = ", f1, ", \nAUC_xgb = ", Auc_xgb)


XGBoost by Hold-Out Evaluation: 
Accuracy =  0.89868 , 
F1 =  0.4733183053489793 , 
AUC_xgb =  0.5


## Conclusions:

### Best Performers:
- **Random Forest** consistently performs very well with perfect or near-perfect scores (Accuracy = 1.0, F1 = 1.0, AUC = 1.0).
- **Gradient Boosting Classifier** also performs exceptionally well, with high scores across the board (Accuracy = 1.0, F1 = 1.0, AUC = 1.0).
- **AdaBoosting** shows strong performance with high scores (Accuracy = 0.89868, F1 = 0.47332, AUC = 0.5).
- **XGBoost** also performs well with similar scores to AdaBoosting (Accuracy = 0.89868, F1 = 0.47332, AUC = 0.5).

### Linear SVM:
- **Linear SVM** has a lower accuracy compared to other models, especially the random forest, gradient boosting, AdaBoosting, and XGBoost (Accuracy = 0.76064).
- It has a higher F1 score compared to SVMs with other kernels but still lower than the best performers (F1 = 0.62436).
- AUC is highest among SVM models (AUC = 0.77201).

### SVM with Polynomial, RBF, and Sigmoid Kernels:
- These **SVM models** have identical scores, indicating they are likely not performing well on this dataset (Accuracy = 0.89928, F1 = 0.47348, AUC = 0.5).
- Their accuracy, F1, and AUC scores are consistently lower compared to the best performers.

### Logistic Regression:
- **Logistic Regression** performs reasonably well but is outperformed by Random Forest, Gradient Boosting Classifier, AdaBoosting, and XGBoost (Accuracy = 0.9047, F1 = 0.6468, AUC = 0.6145).

### Bagging with Bernoulli Naive Bayes:
- **Bagging with Bernoulli Naive Bayes** has lower scores across the board (Accuracy = 0.89928, F1 = 0.47348, AUC = 0.5).

Based on the provided metrics, the **Random Forest, Gradient Boosting Classifier, AdaBoosting**, and **XGBoost** models stand out as the best performers, achieving near-perfect scores across all metrics. Linear SVM, while having the highest AUC among SVM models, falls short compared to these top performers.
 these top performers.
 these top performers.
0, F1 = 1.0, AUC = 1.0).
ts of each model's performance metrics.
across the board.
